### PR TITLE
Add ELU, PReLU, ThresholdedReLU layers

### PR DESCRIPTION
**Pull request by plavin**
This is based off of the Keras documentation [here](https://keras.io/layers/advanced-activations/) The PReLU layer in Keras supports more features that we can discuss adding in the future. (e.g. weight initialization function, regularization).

### CODE REVIEW

File: src/nn/Modules/Activations.cpp [hunk 1]
```diff
@@ -46,5 +46,48 @@ namespace af
...
+        PReLU::PReLU(const Variable &w) :
+            Module({w})
+        {
+        }
+
+        Variable PReLU::forward(const Variable &input)
+        {
+            auto tmp = max(input, 0.0);
+            auto res = expandAs(m_parameters[0],tmp) * tmp;
+            //TODO: Determine if doing the max after the mul is preferable
```
Comment (pavanky) [inline src/nn/Modules/Activations.cpp]: That would be two different behaviors. max after multiply will have a different behavior when values of `input` and `m_parameters` are both negative. Can you double check what the exact behavior should be.

File: src/nn/Modules/Activations.cpp [hunk 2]
```diff
@@ -46,5 +46,48 @@ namespace af
...
+        ThresholdedReLU::ThresholdedReLU(double threshold) :
+            m_threshold(threshold)
+        {
+        }
+
+        Variable ThresholdedReLU::forward(const Variable &input)
+        {
+            auto mask = input > m_threshold;
+            return input * mask;
+        }
```
Comment (pavanky) [inline src/nn/Modules/Activations.cpp]: This should probably just be renamed to `Threshold` ?

Comment (FloopCZ) [reply src/nn/Modules/Activations.cpp]: I believe `Threshold` may be misleading, since threshold usually denotes a function R -> {0,1} (e.g., `return input > m_threshold;`). `ThresholdedReLU` or `ThresholdReLU` seem a better fit to me.

Comment (pavanky) [reply src/nn/Modules/Activations.cpp]: `ThresholdReLU` sounds better. Thanks @FloopCZ

File: src/autograd/Functions.cpp [hunk 3]
```diff
@@ -113,6 +129,18 @@ namespace af {
...
+          auto mask = lhs < rhs;
+          auto result = min(lhs.array(), rhs.array());
+
+          auto grad_func = [](std::vector<Variable> &inputs, const Variable &grad_output) {
+            inputs[0].addGrad( inputs[2] * grad_output);
+            inputs[1].addGrad(!inputs[2] * grad_output);
+          };
+          return Variable(result, {lhs, rhs, mask}, grad_func);
+        }
+
```
Comment (FloopCZ) [inline src/autograd/Functions.cpp]: This function should be indented by four spaces.

File: src/nn/Modules/Activations.cpp [hunk 4]
```diff
@@ -74,18 +71,18 @@ namespace af
 
         Variable ELU::forward(const Variable &input)
         {
-            auto res = max(input, m_alpha * (exp(input) - 1));
-            return res;
+            auto mask = input >= 0.0;
+            return (mask * input) + (!mask * m_alpha * exp(input));
```
Comment (pavanky) [inline src/nn/Modules/Activations.cpp]: Shouldn't this be `(!mask * m_alpha * (exp(input) - 1))` ?
https://keras.io/layers/advanced-activations/#elu

Comment (plavin) [reply src/nn/Modules/Activations.cpp]: dangit...
